### PR TITLE
[X86][CFE] Support EGPR in GCCRegNames.

### PR DESCRIPTION
**Pull request by FreddyLeaf**

### CODE REVIEW

File: clang/lib/Basic/Targets/X86.cpp [hunk 1]
```diff
@@ -1763,10 +1823,14 @@ void X86TargetInfo::fillValidTuneCPUList(SmallVectorImpl<StringRef> &Values) con
 }
 
 ArrayRef<const char *> X86TargetInfo::getGCCRegNames() const {
+  if (HasEGPR)
+    return llvm::ArrayRef(ExtendedGCCRegNames);
   return llvm::ArrayRef(GCCRegNames);
```
Comment (KanRobert) [inline clang/lib/Basic/Targets/X86.cpp:1828]: Shouldn't we extend GCCRegNames directly regardless of the target feature?

Comment (FreddyLeaf) [reply clang/lib/Basic/Targets/X86.cpp:1828]: I noticed gcc did so. it reports error without -mapxf when compiling `r16`. reasonable to me.

Comment (KanRobert) [reply clang/lib/Basic/Targets/X86.cpp:1828]: It's a gap. https://www.godbolt.org/z/xx8nrMdYW
But please keep the behavior consistent inside clang

File: clang/test/CodeGen/X86/register_asm.c [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-unknown %s -o /dev/null
+// RUN: %clang_cc1 -triple x86_64-unknown-unknown -target-feature +egpr %s -o /dev/null
+
+int foo(void) {
+  register int a __asm__("ebx");
+#ifdef __EGPR__
```
Comment (KanRobert) [inline clang/test/CodeGen/X86/register_asm.c:6]: I believe this test can be simplified and merged into an existing test.

Comment (FreddyLeaf) [reply clang/test/CodeGen/X86/register_asm.c:6]: which one? I can only found relate one: clang/test/CodeGen/2004-03-16-AsmRegisterCrash.c but with a bad location and file name.

File: clang/lib/Basic/Targets/X86.cpp (hunk 1 above)
Comment (FreddyLeaf) [reply clang/lib/Basic/Targets/X86.cpp]: Got it. Addressed in [22a1795](https://github.com/llvm/llvm-project/pull/91323/commits/22a17952999aa2b10b8a71c71f7607e37290b973)

File: clang/lib/Basic/Targets/X86.cpp [hunk 3]
```diff
@@ -83,8 +85,23 @@ const TargetInfo::AddlRegName AddlRegNames[] = {
     {{"r13d", "r13w", "r13b"}, 43},
     {{"r14d", "r14w", "r14b"}, 44},
     {{"r15d", "r15w", "r15b"}, 45},
+    {{"r16d", "r16w", "r16b"}, 165},
```
Comment (KanRobert) [inline clang/lib/Basic/Targets/X86.cpp:88]: Where does the 165 come from?

Comment (FreddyLeaf) [reply clang/lib/Basic/Targets/X86.cpp:88]: it's the index of "r16" in GCCRegNames

File: clang/test/CodeGen/X86/register_asm.c (hunk 2 above)
Comment (KanRobert) [reply clang/test/CodeGen/X86/register_asm.c]: We have

./clang/test/CodeGen/M68k/inline-asm-gcc-regs.c
./clang/test/CodeGen/LoongArch/inline-asm-gcc-regs.c


I suggest splitting ./clang/test/CodeGen/2004-03-16-AsmRegisterCrash.c into 

```
./clang/test/CodeGen/ARM/inline-asm-gcc-regs.c
./clang/test/CodeGen/X86/inline-asm-gcc-regs.c
```
and add your test into the second one.

Comment (FreddyLeaf) [reply clang/test/CodeGen/X86/register_asm.c]: what about rename this test into clang/test/CodeGen/X86/inline-asm-gcc-regs.c?

Comment (FreddyLeaf) [reply clang/test/CodeGen/X86/register_asm.c]: there's no clang/test/CodeGen/ARM/

Comment (KanRobert) [reply clang/test/CodeGen/X86/register_asm.c]: Create one.

Comment (FreddyLeaf) [reply clang/test/CodeGen/X86/register_asm.c]: but there are many clang/test/CodeGen/aarch64-*/ And I don't know which one to put it in.

Comment (KanRobert) [reply clang/test/CodeGen/X86/register_asm.c]: bash$ ls -d llvm/test/CodeGen/*/

llvm/test/CodeGen/ARM/ 
llvm/test/CodeGen/AArch64/

ARM use a separate dir from AArch64. So create clang/test/CodeGen/ARM/ and put test in it.

Comment (phoebewang) [reply clang/test/CodeGen/X86/register_asm.c]: Why just use `#ifdef __arm__` to distinguish ARM from X86?

Comment (FreddyLeaf) [reply clang/test/CodeGen/X86/register_asm.c]: > bash$ ls -d llvm/test/CodeGen/*/
> 
> llvm/test/CodeGen/ARM/ llvm/test/CodeGen/AArch64/
> 
> ARM use a separate dir from AArch64. So create clang/test/CodeGen/ARM/ and put test in it.

There's no even AArch64 under clang/test/CodeGen/. ARM's clang codegen lit tests are not organized well if you look into clang/test/CodeGen/. It'd better not create a location if it won't be used in the future.

File: clang/test/CodeGen/X86/register_asm.c
Comment (KanRobert) [inline clang/test/CodeGen/X86/register_asm.c:1]: Rename it to inline-asm-gcc-regs.c?

Comment (FreddyLeaf) [reply clang/test/CodeGen/X86/register_asm.c:1]: [a627464](https://github.com/llvm/llvm-project/pull/91323/commits/a627464ef0ed4e999dc52b6b2d445548b559fe8b)

**Review by KanRobert — APPROVED** [review]
LGTM

**Pull request merged**